testrail/api.py

```diff
@@ -1,17 +1,76 @@
 from __future__ import division
 
+import os
 import collections
+from builtins import dict
 from datetime import datetime, timedelta
-import os
 
-import requests
 import yaml
+import requests
 
 from testrail.helper import TestRailError
 
 nested_dict = lambda: collections.defaultdict(nested_dict)
 
 
+class UpdateCache(object):
+    """ Decorator class for updating API cache
+    """
+    def __init__(self, cache):
+        self.cache = cache
+
+    def __call__(self, f):
+        def wrapped_f(*args, **kwargs):
+            resp = f(*args, **kwargs)
+            if isinstance(resp, dict) and not resp:
+                # Empty dict, indicating something at args[-1] was deleted.
+                self._delete_from_cache(args[-1])
+            else:
+                # Something must have been added or updated
+                self._update_cache(resp)
+
+            return resp
+        return wrapped_f
+
+    def _delete_from_cache(self, delete_id):
+        ''' Check every dict inside of self.cache for an object with a matching
+            ID
+        '''
+        for project in self.cache.values():
+            obj_list = project['value']
+            for index, obj in enumerate(obj_list):
+                if obj['id'] == delete_id:
+                    obj_list.pop(index)
+                    return
+        else:
+            # If we hit this, it means we went looked at every object in every
+            # cache dict and didn't find a match. Raise an exception
+            exc_msg = "Attempted to delete an item the cache, but could not " +\
+                      "locate item with id {0}".format(delete_id)
+            raise TestRailError(exc_msg)
+
+    def _update_cache(self, update_obj):
```

Comment:
Similarly to delete just looking for confirmation the API call already completed before running update.